src/vs/editor/contrib/stickyScroll/browser/stickyScrollModelProvider.ts

```diff
@@ -0,0 +1,411 @@
+import { StopWatch } from 'vs/base/common/stopwatch';
+import { IFeatureDebounceInformation, ILanguageFeatureDebounceService } from 'vs/editor/common/services/languageFeatureDebounce';
+import { onUnexpectedError } from 'vs/base/common/errors';
+import { TextModel } from 'vs/editor/common/model/textModel';
+import { StickyElement, StickyModel, StickyRange } from 'vs/editor/contrib/stickyScroll/browser/stickyScrollElement';
+import { Iterable } from 'vs/base/common/iterator';
+
+enum ModelProvider {
+	OUTLINE_MODEL = 'outlineModel',
+	FOLDING_PROVIDER_MODEL = 'foldingProviderModel',
+	INDENTATION_MODEL = 'indentationModel'
+}
+
+enum Status {
+	VALID,
+	INVALID,
+	CANCELED
+}
+
+export interface IStickyModelProvider {
+
+	/**
+	 * Method which updates the sticky model
+	 * @param textModel text-model of the editor
+	 * @param textModelVersionId text-model version ID
+	 * @param token cancellation token
+	 * @returns the sticky model
+	 */
+	update(textModel: ITextModel, textModelVersionId: number, token: CancellationToken): Promise<StickyModel | null>;
+}
+
+export class StickyModelProvider implements IStickyModelProvider {
+
+	private _modelProviders: IStickyModelCandidateProvider[] = [];
+	private _modelPromise: CancelablePromise<any | null> | null = null;
+	private _updateScheduler: Delayer<StickyModel | null> | null = null;
+	private readonly _updateDebounceInfo: IFeatureDebounceInformation;
+	private readonly _store: DisposableStore;
+
+	constructor(
+		private _editor: ICodeEditor,
+		@ILanguageFeaturesService _languageFeaturesService: ILanguageFeaturesService,
+		@ILanguageConfigurationService _languageConfigurationService: ILanguageConfigurationService,
+		@ILanguageFeatureDebounceService _languageFeatureDebounceService: ILanguageFeatureDebounceService,
+		defaultModel: string) {
+
+		switch (defaultModel) {
+			case ModelProvider.OUTLINE_MODEL:
+				this._modelProviders.push(new StickyModelFromCandidateOutlineProvider(_languageFeaturesService));
+				this._modelProviders.push(new StickyModelFromCandidateSyntaxFoldingProvider(this._editor, _languageFeaturesService));
+				this._modelProviders.push(new StickyModelFromCandidateIndentationFoldingProvider(this._editor, _languageConfigurationService));
+				break;
+			case ModelProvider.FOLDING_PROVIDER_MODEL:
+				this._modelProviders.push(new StickyModelFromCandidateSyntaxFoldingProvider(this._editor, _languageFeaturesService));
+				this._modelProviders.push(new StickyModelFromCandidateIndentationFoldingProvider(this._editor, _languageConfigurationService));
+				break;
+			case ModelProvider.INDENTATION_MODEL:
+				this._modelProviders.push(new StickyModelFromCandidateIndentationFoldingProvider(this._editor, _languageConfigurationService));
+				break;
+		}
+
+		this._updateDebounceInfo = _languageFeatureDebounceService.for(_languageFeaturesService.foldingRangeProvider, 'Sticky Scroll Folding', { min: 200 });
```

Comment:
`_languageFeaturesService.foldingRangeProvider` needs to match teh provdier used.
So the `_updateDebounceInfo` needs to go inside `modelProvider.computeStickyModel`